ext/opentelemetry-ext-wsgi/src/opentelemetry/ext/wsgi/__init__.py

```diff
@@ -46,14 +47,43 @@ def _add_request_attributes(span, environ):
         span.set_attribute("component", "http")
         span.set_attribute("http.method", environ["REQUEST_METHOD"])
 
-        host = environ.get("HTTP_HOST") or environ["SERVER_NAME"]
+        host = environ.get("HTTP_HOST")
+        if not host:
+            host = environ["SERVER_NAME"]
+            if environ["wsgi.url_scheme"] == "https":
+                if environ.get("SERVER_PORT", "443") != "443":
+                    host += ":" + environ["SERVER_PORT"]
+            elif environ.get("SERVER_PORT", "80") != "80":
+                host += ":" + environ["SERVER_PORT"]
+
+        # NOTE: Nonstandard, may (not) include port
         span.set_attribute("http.host", host)
 
-        url = (
-            environ.get("REQUEST_URI")
-            or environ.get("RAW_URI")
-            or wsgiref_util.request_uri(environ, include_query=False)
-        )
+        url = environ.get("REQUEST_URI") or environ.get("RAW_URI")
+
+        if url:  # We got something, but is absolute?
+            # The simplistic ``"://" in url` is not sufficient,
+            # as that could be contained in the query string.
+            try:
+                urlparts = urlparse(url)
+            except Exception:  # pylint:disable=broad-except
+                url = wsgiref_util.request_uri(environ)
+            else:
+                if url.startswith("//"):  # Scheme-relative URL
+                    url = url[2:]
+                if (
+                    not urlparts.netloc and urlparts.scheme
+                ):  # E.g., "http:///?"
+                    scheme, path = url.split("://", 1)
+                    url = scheme + "://" + host + path
+                elif not urlparts.netloc or not urlparts.scheme:
+                    scheme = environ["wsgi.url_scheme"] + "://"
+                    if not urlparts.netloc:
+                        url = host + url
+                    url = scheme + url
```

Comment:
Also, are these cases that we have to cover?
```
        self.environ["RAW_URI"] = "/?"
        self.environ["RAW_URI"] = "http:///?"
        self.environ["RAW_URI"] = "http://?"
```

These cases seem to add a lot of complexity. I'm curious to know which WSGI servers write these `RAW_URI` strings?

Reply:
Looking more at this, in order to be completely spec compliant it feels like calling `url = wsgiref_util.request_uri(environ)` on every request would probably be easiest rather than playing with `RAW_URI` and `REQUEST_URI` as an optimization. What are your thoughts on this?

Reply:
On the one hand, I think using RAW_URI et al is valuable because it catches things like `http://myhost/?` vs `http://myhost/` that should not matter but could. And since one of the prime use cases of tracing is debugging, I think exactly these edge cases are interesting to handle. But on the other hand, I might have gone overboard with my handling of malformed/very unusual raw URIs. Maybe there is a middle ground to aim for.

I also don't think the calculation for host is particularly expensive and it could probably even be optimized. I'd wait until the data formats are reworked (this is something that will most likely happen).

Reply:
Not sure if we can get something like `http://foobar.org/abc#CrazyFragment` and how do we plan to handle it.
I swear that I've seen something like this from server side (which means both client and server are super buggy).

Reply:
If we get that all in RAW_URL, we'll set it to the URL as-is. If we get invalid data, we are not in the business to normalize it, instead we should make it visible.

Reply:
I simplified it:
* If the raw URL starts with `/` we assume it is relative and prepend scheme://host
* Otherwise, if the url starts with "scheme:", we use it as-is
* Otherwise we assume a bogus value and fall back to wsgiref.util.request_uri.